### PR TITLE
feat: add support for detached signatures and document canonicalization

### PR DESCRIPTION
**Pull request by ntn-x2**
Fixes https://github.com/KILTprotocol/ticket/issues/2670.

This PR introduces the following changes:

* Add support for OPTIONAL signatures for accounts linked to a specific asset
* [BREAKING] Require the document to be canonicalized following [RFC 8785](https://tools.ietf.org/html/rfc8785) before being hashed
* Minor editorial changes

## Checklist

- [ ] Remove TODOs
- [ ] Update documents, hashes, and IPFS CID once the final document structure is agreed upon

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -20,76 +22,174 @@ This document defines an extension to the service types supported in the [DID Co
...
+* `ecdsa-secp256k1`: the EDDSA digital signature scheme based on the secp256k1 curve. They are typically identified as `ecdsa` in the Polkadot ecosystem and are the default scheme for Ethereum and Bitcoin accounts.
+
+#### Hashing Schemes
+
+* `blake2b-256`: Blake2b hash with 256 bit output. They are the most common choice for Polkadot networks.
+* `keccak-256`: Hashing scheme used by Ethereum with 256 bit output.
+
+### Signature Generation and Verification
+
+Wherever possible, accounts exposed as part of this spec SHOULD include the optional `proof` property that provides evidence of control of such an account by the DID subject.
```
Comment (ERussel) [inline README.md:165]: The only point I see having proof in the JSON is to ensure that a user has access to the account that is being provided. But I don't think that it is necessary to include the signature in the json, dapp can just check it by itself and notify the user on ui/ux level.

File: README.md [hunk 2]
```diff
@@ -20,76 +22,174 @@ This document defines an extension to the service types supported in the [DID Co
...
+
+const doc = `
+{
+  "polkadot:b0a8d493285c2df73290dfb7e61f870f/slip44:434": {
+    "EJDj2GKnx89HTzUkGW8Rk9RoYUmAJHPM8aacWFp3fi1gYUQ": {
+      "description": "Personal account"
+    }
+  },
+  "polkadot:411f057b9107718c9624d6aa4a3f23c1/slip44:2086": {
+    "4nvZhWv71x8reD9gq7BUGYQQVvTiThnLpTTanyru9XckaeWa": {
```
Comment (ERussel) [inline README.md:116]: Why it is required to change the format of the JSON breaking backward compatibility?

Comment (ntn-x2) [reply README.md:116]: Unfortunately it's because arrays are not sorted as part of the canonicalisation process: https://www.rfc-editor.org/rfc/rfc8785#section-3.2.3-5.2.

The JSON spec does not have a concept for unordered set, hence arrays are considered ordered data structures, and its elements not re-arranged. That's why the change, unfortunately.

File: README.md (hunk 1 above)
Comment (ntn-x2) [reply README.md:165]: Agree. Will remove that from the spec. It makes everyone's life so much easier.

File: README.md (hunk 2 above)
Comment (ERussel) [reply README.md:116]: Got it. But do we really want to treat a list of accounts per chain as an unordered set? Because UI will usually display accounts as the list, a user may want to put more important accounts to the top.

Comment (ntn-x2) [reply README.md:116]: That's a conversation we might want to have. For instance URLs for a service endpoint are typically treated as an unordered set. In this case, we do have an optional `description` field which would let the user specify additional details. I would see this field as being present most of the time, and each address being dedicated to something different, potentially. I do not see the case why a user would expose two addresses for the same purpose, e.g., receiving simple transfers, where one address would be more important than the other. Could you see a case like that? Or have you seen such a case?

TL;DR: I think multiple addresses for the same asset class will usually be used for different purposes, which makes them not naturally "sortable" between them anyway. If two addresses are present for the very same purpose, I would like to know why a user might/could prefer one over the other.

**Comment by ntn-x2** [comment]
Superseded by https://github.com/KILTprotocol/spec-KiltTransferAssetRecipient/pull/3.
